src/wex/Engine.cpp

```diff
@@ -0,0 +1,55 @@
+#include "SFML/System/Clock.hpp"
+#include "SFML/Window/Keyboard.hpp"
+#include <SFML/Graphics.hpp>
+#include <SFML/Window.hpp>
+
+#include <wex/Engine.hpp>
+
+#include <cassert>
+#include <iostream>
+
+namespace wex {
+
+Engine::Status Engine::run() {
+	mGame->init();
+	loop();
+	return Status::ok;
+}
+
+void Engine::handleEvents(sf::RenderWindow& window) {
+	sf::Event event;
+	while (window.pollEvent(event)) {
+		switch (event.type) {
+		case sf::Event::Closed: window.close(); break;
+		case sf::Event::KeyPressed: std::cout << "key pressed\n"; break;
+		default: break;
+		}
+	}
+}
+
+void Engine::loop() {
+	const double msPerFrame = 1000.0 / mGraphics.mFps;
+
+	sf::Clock clock;
+	double lag = 0.0;
+
+	auto& window = mGraphics.mWindow;
+	while (mGraphics.mWindow.isOpen()) {
+		const int dt = clock.restart().asMilliseconds();
+		lag += dt;
+
+		handleEvents(window);
+
+		while (lag >= msPerFrame) {
+			mGame->update(dt);
```

Comment:
Sure you want to go with fixed time step? Such loops are harder to debug, since you may not "see" every frame

Reply:
So are you saying it's a better idea to let the user decide the time-step in the `update`?
Or give them control of the `loop` altogether?

Reply:
Not exactly; I'm saying in general it's easier to debug variable time step core loops, which is why most third party engines use that approach too. Both Unity and Unreal's main update loop is variable, and they provide a fixed update version for physics etc. As such that provision isn't even necessary, a variable update function can perform N fixed time slice ticks by itself.

User controlling game loop will free you of having to take such decisions, but will also make usage of the engine a bit harder. I use this approach, BTW.